kernel/rtlil.h

```diff
@@ -1222,13 +1222,10 @@ struct RTLIL::Module : public RTLIL::AttrObject
 	RTLIL::Cell* addFf    (RTLIL::IdString name, const RTLIL::SigSpec &sig_d, const RTLIL::SigSpec &sig_q, const std::string &src = "");
 	RTLIL::Cell* addDff   (RTLIL::IdString name, const RTLIL::SigSpec &sig_clk, const RTLIL::SigSpec &sig_d,   const RTLIL::SigSpec &sig_q, bool clk_polarity = true, const std::string &src = "");
 	RTLIL::Cell* addDffe  (RTLIL::IdString name, const RTLIL::SigSpec &sig_clk, const RTLIL::SigSpec &sig_en,  const RTLIL::SigSpec &sig_d, const RTLIL::SigSpec &sig_q, bool clk_polarity = true, bool en_polarity = true, const std::string &src = "");
-	RTLIL::Cell* addDffsr (RTLIL::IdString name, const RTLIL::SigSpec &sig_clk, const RTLIL::SigSpec &sig_set, const RTLIL::SigSpec &sig_clr,
-			RTLIL::SigSpec sig_d, const RTLIL::SigSpec &sig_q, bool clk_polarity = true, bool set_polarity = true, bool clr_polarity = true, const std::string &src = "");
-	RTLIL::Cell* addAdff (RTLIL::IdString name, const RTLIL::SigSpec &sig_clk, const RTLIL::SigSpec &sig_arst, const RTLIL::SigSpec &sig_d, const RTLIL::SigSpec &sig_q,
-			RTLIL::Const arst_value, bool clk_polarity = true, bool arst_polarity = true, const std::string &src = "");
```

Comment:
IMO the parser in pyosys should be fixed rather than the code in core Yosys files be modified to work around its deficiencies.

Reply:
Agreed. There were more issues in the parser though that I didn't want to open that box.
It might be worth evaluating if the header parsing business could be circumvented.

Reply:
I'm wondering if SWIG could be used instead of boost::python; as a side effect the API wouldn't be just exported to Python but to many other languages as well.

Reply:
I've used SWIG in the past, it used to be problematic with nested user-defined object structures and ownerships, might have gotten better. Maybe a wrapper class would solve some problems, but also introduce an (internal) API change, possibly. The other option I've found quite effective is Cython (python-wrapping only, obviously). This allows some tight interaction with the C++ code while taking care of the reference counting and object ownerships.
I have found little references on the pyosys API and its testing so far, maybe someone from the maintainer's side can share more insight?

Reply:
> it used to be problematic with nested user-defined object structures and ownerships, might have gotten better.

AFAICT the ownership management in pyosys is a rather invasive hack that relies on `get_all_*()` functions. SWIG can't possibly be worse because you can just keep using that.

Reply:
I agree that it would be better to fix the parser. I am not sure but I thought multi-line definitions were supported.

> AFAICT the ownership management in pyosys is a rather invasive hack that relies on get_all_*() functions. SWIG can't possibly be worse because you can just keep using that.

True. I never worked with SWIG and have no idea how much work it would be to get it up and running. I expect the result to be much better though as the current parser only supports (most of) the c++ style that was used when I wrote it.

Reply:
> I have found little references on the pyosys API and its testing so far, maybe someone from the maintainer's side can share more insight?

The pyosys API was developed as a student project by @btut. It doesn't have as many users since it isn't enabled by default, so it's relatively rarely touched, and I believe none of the maintainers has worked with it enough to develop a thorough understanding of the code yet. As far as I know, there are no tests, at least in `make test` (which is what runs in our CI).

Reply:
I do know of some people that are using it, but there are probably not many out there.

I don't think the userbase is important here though. Ideally, the interfaces would stay the same no matter which tool (parser+boostpython vs SWIG) does the translation.

If the parser stays, adding some tests would definitely be a good idea. What do you have in mind? test the build, check if a set of functions are exposed to python, and then do some basic interaction with Yosys?

Reply:
Oh no, the userbase comment wasn't meant about usefulness, just about testing. Since our test suite still has a lot of deficiencies, we rely to a significant extent on someone yelling if we break something 😅 Because pyosys isn't used by as many people, it takes longer for someone to notice problems. (And we also look at the code less often than other parts of yosys since there are fewer issues/PRs about it.)

A few unit tests would definitely be an improvement. The tests included in ``make test`` should be short (the whole thing has to fit in travis' 1 hour limit), so just exercising the basic functions on a small design sounds good. We could also add a larger integration test, perhaps trying to synthesize a more complex design, to yosys-tests (which we are in the process of setting up as an additional, longer-running CI task).

Reply:
> I don't think the userbase is important here though. Ideally, the interfaces would stay the same no matter which tool (parser+boostpython vs SWIG) does the translation.

For me, API changes are no problem, as I'm wrapping things up to a certain point.
The project (direct synthesis from Python/MyHDL) is here: https://github.com/hackfin/myhdl/tree/jupyosys.
The minor issues I had are mostly the ones in #2043, however there are some API specifics that appear 'unpythonic', for instance yosys dictionary types access (where only the full assignment of a dict is supported and insertion by key would silently be ineffective).

> If the parser stays, adding some tests would definitely be a good idea. What do you have in mind?

I keep relying on py.test mostly, you'll find a few scenarios in the above MyHDL fork. There's also a nice integration with jupyter notebooks so you can make sure that documented examples also are up to date (you'll find these in the secondary-pulled 'hacker space' from the jupyosys binder environment)